openshift/machine-config-operator
# MCO-564: Make NodeController aware of BuildController

**- What I did**

This makes NodeController aware of BuildController. Specifically, it will
propagate the final OS image pullspec from the MachineConfigPool annotation
(`machineconfiguration.openshift.io/newestImageEquivalentConfig`) into the
currentImage and desiredImage annotations for each node in the target pool.

**- How to verify it**

I've written a small Go binary that can be used to help aid in these steps: https://github.com/openshift/machine-config-operator/pull/3852. It is not required, but does make the setup / teardown process of this much easier.

1. Bring up an OpenShift cluster.
2. Create a new MachineConfigPool or use an existing MachineConfigPool.
3. Opt a pool into layering by adding the `machineconfiguration.openshift.io/layering-enabled` label onto a MachineConfigPool.
4. Set `machineconfiguration.openshift.io/newestImageEquivalentConfig` to an arbitrary value, e.g., `registry.hostname.com/org/repo:tag-1`.
5. Set the `BuildSuccess` condition on the MachineConfigPool. Note: You may have to use my onclusterbuilds binary found here https://github.com/openshift/machine-config-operator/pull/3852 in order to easily do that: `$ ./onclustertesting set-build-status --pool=<poolname> --type BuildSuccess --status=true`
6. Add a node to a pool by setting the `node-role.kubernetes.io/<poolname>` label on the node.

At this point, you should see the MachineConfigPool transition into the `Updating` status. You should also see that the node you've added to the pool has the annotation `machineconfiguration.openshift.io/desiredImage: "registry.hostname.com/org/repo:tag-1"` present. Once the node gets the `machineconfiguration.openshift.io/currentImage: "registry.hostname.com/org/repo:tag-1"` annotation it will move onto the next node in the pool until no more nodes require an update. Adding the `currentImage` annotation will become the job of the MCD once https://github.com/openshift/machine-config-operator/pull/3848 lands. Until that PR lands, you will have to add the `currentImage` annotation to the node to simulate the MCD's behavior.

**- Description for the changelog**
Makes NodeController aware of BuildController
